File: src/service/Feed.service.ts
import dayjs from 'dayjs';

import { feedApi } from '@/api/feed/Feed.api';

import { AddFeedSourceRequest } from '@/api/feed/types/AddFeedSource.type';
import { FeedArticle } from '@/types/FeedArticle.type';

class FeedService {

    async getFeed(): Promise<Array<FeedArticle> | Error> {
        const getFeedResponse = await feedApi.getFeed();

        if (getFeedResponse instanceof Error)
            return getFeedResponse;

        return getFeedResponse.sources.map(source => source.articles
            .map(article => (<FeedArticle>{
                url: article.url,
                source: source,
                title: article.title,
                summary: article.summary,
                author: article.author,
                publishedAt: dayjs(article.publishedAt ?? article.publishedAtAsString),
            }))
        )
        .flat();
    }

    async addFeedSource(request: AddFeedSourceRequest): Promise<Array<FeedArticle> | Error> {
        const addFeedSourceResponse = await feedApi.addFeedSource(request);

        if (addFeedSourceResponse instanceof Error)
            return addFeedSourceResponse;

        return addFeedSourceResponse.articles.map(x => ({
            url: x.url,
            title: x.title,
            summary: x.summary,
            author: x.author,
            publishedAt: dayjs(x.publishedAt ?? x.publishedAtAsString),
            source: {
                reference: addFeedSourceResponse.reference,
                title: addFeedSourceResponse.title,
                description: addFeedSourceResponse.description,
                rssUrl: addFeedSourceResponse.rssUrl,
                websiteUrl: addFeedSourceResponse.websiteUrl,
            },
        }));
    }

    async deleteFeedSource(reference: string): Promise<void | Error> {
        const deleteFeedSourceResponse = await feedApi.deleteFeedSource(reference);

        if (deleteFeedSourceResponse instanceof Error)
            return deleteFeedSourceResponse;
    }
}

export const feedService = new FeedService();

File: backend/MightyRSS/Api/Feed/FeedService.cs
using Data.Records;
using Data.UoW;
using MightyRSS.Api.Feed.Types;
using MightyRSS.Models.Mappers;
using NetApiLibs.Type;

namespace MightyRSS.Api.Feed;

public interface IFeedService
{
    Task<Result<GetFeedResponse>> GetFeed(UserRecord user, CancellationToken cancellationToken);
}

public sealed class FeedService : IFeedService
{
    private readonly IUnitOfWorkFactory<IMightyUnitOfWork> _mightyUnitOfWorkFactory;

    public FeedService(IUnitOfWorkFactory<IMightyUnitOfWork> mightyUnitOfWorkFactory)
    {
        _mightyUnitOfWorkFactory = mightyUnitOfWorkFactory;
    }

    public async Task<Result<GetFeedResponse>> GetFeed(UserRecord user, CancellationToken cancellationToken)
    {
        using var unitOfWork = _mightyUnitOfWorkFactory.Create(cancellationToken);

        var userFeedSources = await unitOfWork.UserFeedSources.GetFeedSources(user);

        await unitOfWork.Commit();

        return new GetFeedResponse
        {
            Sources = userFeedSources.ConvertAll(userFeedSource => new GetFeedResponse.FeedSourceDetails
            {
                FeedSource = FeedSourceMapper.Map(userFeedSource.FeedSource, userFeedSource),
                Articles = userFeedSource.FeedSource.Articles.ConvertAll(article => new GetFeedResponse.FeedArticle
                {
                    Url = article.Url,
                    Title = article.Title,
                    Summary = article.Summary,
                    Author = article.Author,
                    PublishedAt = article.PublishedAt,
                    PublishedAtAsString = article.PublishedAtAsString
                })
            })
        };
    }
}